### PR TITLE
Add Element Type To Semantic Text Model Settings

### PR DESCRIPTION
**Pull request by Mikep86**
Adds the element type to `semantic_text`'s model settings, which allows it to index byte embeddings (like those generated by Cohere).

```
PUT _inference/text_embedding/cohere-embeddings-byte
{
    "service": "cohere",
    "service_settings": {
        "api_key": <redacted>,
        "model_id": "embed-english-light-v3.0",
        "embedding_type": "byte",
        "similarity": "dot_product"
    }
}

PUT test-index
{
    "mappings": {
        "properties": {
            "infer_field_byte": {
                "type": "semantic_text",
                "inference_id": "cohere-embeddings-byte"
            },
            "non_infer_field": {
                "type": "text"
            }
        }
    }
}

PUT test-index/_doc/doc1
{
    "infer_field_byte": "these are not the droids you're looking for. He's free to go around",
    "non_infer_field": "hello"
}
```

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java
Comment (Mikep86) [inline server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:1]: @benwtrent I had to update `DenseVectorFieldMapper.Builder` to make `elementType` programmatically configurable

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextField.java [hunk 1]
```diff
@@ -64,20 +65,42 @@ public record SemanticTextField(String fieldName, List<String> originalValues, I
...
+            this(
+                model.getTaskType(),
+                model.getServiceSettings().dimensions(),
+                model.getServiceSettings().similarity(),
+                model.getServiceSettings().elementType()
+            );
         }
 
+        // TODO: Remove once tests are updated
         public ModelSettings(TaskType taskType, Integer dimensions, SimilarityMeasure similarity) {
```
Comment (Mikep86) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextField.java:90]: I'll remove this constructor once all tests are updated

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextField.java [hunk 2]
```diff
@@ -112,10 +141,12 @@ private void validate() {
                 case TEXT_EMBEDDING:
                     validateFieldPresent(DIMENSIONS_FIELD, dimensions);
                     validateFieldPresent(SIMILARITY_FIELD, similarity);
+                    // validateFieldPresent(ELEMENT_TYPE_FIELD, elementType);
```
Comment (Mikep86) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextField.java:144]: This check is commented out right now to keep tests passing (for now)

**Comment by Mikep86** [comment]
@benwtrent Any updates required to the `semantic` query to get it to work correctly with byte embeddings? In particular, I'm looking here: https://github.com/elastic/elasticsearch/blob/ab5b9717c42dd54467204c77958b07844b5dd7a0/x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java#L417

That gets the embedding as a float, even when we're using byte embeddings. Which doesn't seem right...

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:1]: Are all these changes needed? Isn't it enough to add a setter for `public Builder elementType(ElementType elementType)`?  I don't understand if the other changes are for cleanup or if they fix a real issue when elementType is not by the parser.

Comment (Mikep86) [reply server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:1]: Default index settings are dependent on the element type, so those need to change (and be validated) when you change the element type.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextField.java (hunk 2 above)
Comment (jimczi) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextField.java:144]: The element type is not mandatory in the service settings so I think we need to handle the case where the element type is null. In such case we should just consider that the default is float.

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:1]: Hmm, can't we just call `FieldMapper.Builder#validate` in the build method? Looks like something that every field mapper builder should do. Currently the validation is done only after parsing and merging so I think you found an interesting bug.

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java [hunk 3]
```diff
@@ -242,6 +231,17 @@ public Builder dimensions(int dimensions) {
             return this;
         }
 
+        public Builder elementType(ElementType elementType) {
+            this.elementType.setValue(elementType);
+
+            if (this.indexOptions.isConfigured() == false) {
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:237]: The two are indeed intertwined.

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java
Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:1]: We should have a builder validate, but the serialization checks are also necessary as I think those are used when merging mapping updates, not just on initial buid.

**Comment by benwtrent** [comment]
> That gets the embedding as a float, even when we're using byte embeddings. Which doesn't seem right...

I do not know if the semantic query needs updating or not. But it should handle `byte[]` and `float[]`.

As for dynamically transforming from byte[] to float[], that is ok as long as the indexed element type is `float`, however, if the indexed element type is `byte`, we should transform in the opposite direction, and throw if the values provided are not `byte` (i.e. whole numbers between -128..127).

**Comment by Mikep86** [comment]
Dug deeper and found how `KnnVectorQueryBuilder` converts the `float[]` embeddings to `byte[]`. See here: https://github.com/elastic/elasticsearch/blob/ab5b9717c42dd54467204c77958b07844b5dd7a0/server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java#L1540-L1543

The key thing I was missing was [VectorData.asByteVector](https://github.com/elastic/elasticsearch/blob/d471ccb5bb4e5226f843ae688b0b2b205a842131/server/src/main/java/org/elasticsearch/search/vectors/VectorData.java#L50), which handles checking if the `float` values can be transformed to `byte` and then performing the transformation.

Knowing this, it should be fine to get all text embedding query inference results as `float[]`; this logic will transform it to `byte[]` while creating the Lucene query.

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java
Comment (Mikep86) [reply server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:1]: Added a call to `validate()` in the build method, which should cover the validation we need to perform. I didn't find any changes necessary to the serialization checks because those are invoked when the mapper is [written to XContent](https://github.com/elastic/elasticsearch/blob/1080425a65d826df1fae1a45b15fd163c49ff172/server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java#L896), which is independent of validation.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextField.java (hunk 2 above)
Comment (Mikep86) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextField.java]: [Service settings' element type](https://github.com/elastic/elasticsearch/blob/1a7c46a19778ea0f8d74e29da4a56ccd6bddc1a4/server/src/main/java/org/elasticsearch/inference/ServiceSettings.java#L41) is set to a non-null value for every text embedding model, so it should be fine to enforce this check.

File: x-pack/plugin/inference/qa/test-service-plugin/src/main/java/org/elasticsearch/xpack/inference/mock/TestDenseInferenceServiceExtension.java [hunk 4]
```diff
@@ -138,41 +140,78 @@ public void chunkedInfer(
...
         private List<ChunkedInferenceServiceResults> makeChunkedResults(List<String> input, int dimensions) {
-            var chunks = new ArrayList<InferenceTextEmbeddingFloatResults.InferenceFloatEmbedding>();
-            for (int i = 0; i < input.size(); i++) {
-                float[] embedding = generateEmbedding(input.get(i), dimensions);
-                chunks.add(new InferenceTextEmbeddingFloatResults.InferenceFloatEmbedding(embedding));
-            }
-
-            return InferenceChunkedTextEmbeddingFloatResults.listOf(input, new InferenceTextEmbeddingFloatResults(chunks));
+            InferenceTextEmbeddingFloatResults nonChunkedResults = makeResults(input, dimensions);
+            return InferenceChunkedTextEmbeddingFloatResults.listOf(input, nonChunkedResults);
```
Comment (Mikep86) [inline x-pack/plugin/inference/qa/test-service-plugin/src/main/java/org/elasticsearch/xpack/inference/mock/TestDenseInferenceServiceExtension.java:152]: Simplified the logic here. `makeChunkedResults` does essentially the same thing as `makeResults`, except that the former assigns an embedding to each input.

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java [hunk 5]
```diff
@@ -242,8 +242,16 @@ public Builder dimensions(int dimensions) {
...
+        public Builder elementType(ElementType elementType) {
+            this.elementType.setValue(elementType);
+            return this;
+        }
+
         @Override
         public DenseVectorFieldMapper build(MapperBuilderContext context) {
+            // Validate again here because the dimensions or element type could have been set programmatically,
+            // which affects index option validity
+            validate();
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:254]: It would be good to make this call non optional and called automatically for all fields before we build the mapper.
Would deserve a separate pr though so just adding it here to not forget.

File: server/src/test/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapperTests.java [hunk 6]
```diff
@@ -755,6 +756,20 @@ public void testAddDocumentsToIndexBefore_V_7_5_0() throws Exception {
...
+
+        // Build a dense vector field mapper with float element type, which will trigger int8 HNSW index options
+        DenseVectorFieldMapper mapper = new DenseVectorFieldMapper.Builder("test", IndexVersion.current()).elementType(ElementType.FLOAT)
+            .build(context);
+
+        // Change the element type to byte, which is incompatible with int8 HNSW index options
+        DenseVectorFieldMapper.Builder builder = (DenseVectorFieldMapper.Builder) mapper.getMergeBuilder();
+        builder.elementType(ElementType.BYTE);
+        IllegalArgumentException e = expectThrows(IllegalArgumentException.class, () -> builder.build(context));
+        assertThat(e.getMessage(), containsString("[element_type] cannot be [byte] when using index type [int8_hnsw]"));
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapperTests.java:770]: Can we also test when the failure is thrown from the build method? That should happen if an invalid index option is set after modifying the element type?

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextField.java [hunk 7]
```diff
@@ -112,10 +136,12 @@ private void validate() {
                 case TEXT_EMBEDDING:
                     validateFieldPresent(DIMENSIONS_FIELD, dimensions);
                     validateFieldPresent(SIMILARITY_FIELD, similarity);
+                    validateFieldPresent(ELEMENT_TYPE_FIELD, elementType);
```
Comment (jimczi) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextField.java:139]: We need to make sure that this PR is part of the first release for semantic text, otherwise some already indexed documents might miss this new mandatory type. Is it part of the plan?

**Review by jimczi — COMMENTED** [review]
I left some questions but it looks almost ready!

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java (hunk 5 above)
Comment (Mikep86) [reply server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:254]: ++. Let's make sure to create an issue for this.

File: server/src/test/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapperTests.java (hunk 6 above)
Comment (Mikep86) [reply server/src/test/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapperTests.java:770]: This test already covers when an `IllegalArgumentException` is thrown by the `build` method. Currently, there is no programmatic way to set the index options (only through parsing), so we can't change those like we do with the element type in this test. In fact, that's why the test is constructed the way it is; we need to first set the element type to `FLOAT` to create the desired default index options, then we change the element type to `BYTE` to create the incompatibility.

I thought about adding a way to programmatically set the index options, but that didn't seem appropriate when we would only use it in tests.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextField.java (hunk 7 above)
Comment (Mikep86) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextField.java:139]: Exactly. Without this PR, Cohere byte embeddings don't work, so I consider this change essential for the first release.

**Comment by Mikep86** [comment]
@elasticsearchmachine run elasticsearch-ci/part-1

**Review by jimczi — APPROVED** [review]
LGTM

**Review by davidkyle — APPROVED** [review]
LGTM
